### PR TITLE
Feature/issue 226 grade update tests

### LINKED ISSUE #226: Test for updating student grades
Issue #26 / Pull Request #208 (see 7db5367) added a grade change form on student userview pages. Now admins can go to this page (which is much easier to get to than the studentinfo page in the admin panel) and change the grade very easily. Write a test to make sure that this works correctly.

### PR DESCRIPTION
**Pull request by pushkarkumar06**
## Description

Added comprehensive automated tests for the student grade update functionality introduced in PR #208 (Issue #26). This resolves Issue #226 by ensuring the grade update feature works correctly, securely, and handles edge cases gracefully.

Also fixed a bug in `set_student_grad_year()` method that crashed when given invalid input.

## Related Issue

Closes #226
Related: PR #208, Issue #26

## Type of Change

- [x] New feature
- [x] Bug fix
- [x] Test coverage

## Testing

- [x] Existing tests pass
- [x] New tests added (6 total)
- [x] Manually verified

## Test Results

All 6 tests pass:
```
test_admin_can_update_grade ... ok
test_change_grade_form_in_context ... ok
test_invalid_graduation_year_does_not_change_grade ... ok
test_non_admin_cannot_update_grade ... ok
test_student_cannot_update_own_grade ... ok
test_unauthenticated_cannot_update_grade ... ok

Ran 6 tests in 18.104s - OK
```

## AI Disclosure

This pull request was developed with assistance from AI tools (GitHub Copilot). However, all code implementation, logic, and testing were completely reviewed and verified manually before commit. Every line was checked for correctness, security, and alignment with project standards.

## Checklist

- [x] Self-reviewed the code
- [x] Updated documentation (if needed)
- [x] No new warnings or errors introduced
- [x] All tests passing locally

### CODE REVIEW

File: esp/esp/users/models/__init__.py [hunk 1]
```diff
@@ -1143,8 +1143,14 @@ def set_student_grad_year(self, grad_year):
             profile.save()
 
         #   Update the graduation year.
+        #   Guard against non-numeric input (e.g. a typo in the GET parameter)
+        #   so the view degrades gracefully instead of returning a 500.
+        try:
+            parsed_year = int(grad_year)
```
Comment (SuyashJain17) [inline esp/esp/users/models/__init__.py:1227]: This guard looks good and prevents the ValueError crash. Would it make sense to also add a reasonable range check for graduation years (e.g., prevent very large or negative values), or is that handled elsewhere ?

File: esp/esp/program/modules/tests/studentreg.py [hunk 2]
```diff
@@ -317,3 +318,173 @@ def test_extracosts(self):
...
+        # Minimal program scaffold; we only need students, teachers, and admins.
+        kwargs.update({
+            'num_timeslots': 1, 'timeslot_length': 50, 'timeslot_gap': 10,
+            'num_teachers': 1, 'classes_per_teacher': 1, 'sections_per_class': 1,
+            'num_rooms': 1,
+        })
+        super().setUp(*args, **kwargs)
+
+        # Build StudentInfo / RegistrationProfile rows for every student.
+        self.add_student_profiles()
```
Comment (SuyashJain17) [inline esp/esp/program/modules/tests/studentreg.py:531]: Small question: since `setUp()` calls `add_student_profiles()`, it looks like every test student already has a `StudentInfo`. Because of that, the branch in `set_student_grad_year()` that creates a new `StudentInfo` when it’s missing might not actually be exercised in these tests.

Am I understanding this correctly, or is that case handled elsewhere?

**Comment by pushkarkumar06** [comment]
hey @SuyashJain17  Thanks for the review — both points were valid.

I've now added a **range check for `graduation_year`**, accepting only configured grade-option years, and added tests covering the missing `StudentInfo` path plus out-of-range inputs.

Latest commit: `928172b77`.

**Comment by SuyashJain17** [comment]
Always happy to help review !!

**Comment by pushkarkumar06** [comment]
Hii willgearty , 
I have fixed all the comments given by the copilot , now the PR is ready to review , so you can take a look whenever you get time , and tell me to changes if anything need to be good more.

Thanks for your guidance

**Review by willgearty — APPROVED** [review]
Looks good, thanks!
